src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -115,34 +142,111 @@ internal async Task<IEnumerable<IProjectVersionedValue<RestoreData>>> RestoreAsy
-                return true;
+            try
+            {
+                // Prevent extra restore under some conditions.
+                if (IsProjectConfigurationVersionOutOfDate(value.ConfiguredInputs))
+                {
+                    return false;
+                }
 
-            _latestHash = hash;
+                // Restore service always does work regardless of whether the value we pass 
+                // them to actually contains changes, only nominate if there are any.
+                byte[] hash = RestoreHasher.CalculateHash(restoreInfo!);
 
-            JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
-            {
-                return NominateForRestoreAsync(restoreInfo, _projectAsynchronousTasksService.UnloadCancellationToken);
-            });
+                if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
+                {
+                    SaveNominatedConfiguredVersions(value.ConfiguredInputs);
+                    return true;
+                }
+
+                _latestHash = hash;
+
+                _nugetRestoreStarted = true;
+                JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
+                {
+                    return NominateForRestoreAsync(restoreInfo!, _projectAsynchronousTasksService.UnloadCancellationToken);
+                });
+
+                SaveNominatedConfiguredVersions(value.ConfiguredInputs);
 
-            _projectAsynchronousTasksService.RegisterAsyncTask(joinableTask,
-                                                               ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
-                                                               registerFaultHandler: true);
+                _projectAsynchronousTasksService.RegisterAsyncTask(joinableTask,
+                                                                   ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
+                                                                   registerFaultHandler: true);
 
-            // Prevent overlap until Restore completes
-            bool success = await joinableTask;
+                // Prevent overlap until Restore completes
+                success = await joinableTask;
 
-            HintProjectDependentFile(restoreInfo);
+                lock (SyncObject)
+                {
+                    _nugetRestoreStarted = false;
+                }
+
+                HintProjectDependentFile(restoreInfo!);
+            }
+            catch
+            {
+                success = false;
+            }
+            finally
+            {
+                _nugetRestoreStarted = false;
+            }
 
             return success;
         }
 
+        private bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput>? configuredInputs = null)
```

Comment:
I think the logic here is incorrect, this function is called when we receives package restore data, and our logic is to skip it, if the project has changed to a new version. So the logic should not use anything in _savedNominatedConfiguredVersion at all.

The logic here should be similar to what you did in IsSavedNominationEmptyOrOlder, It is only IsSavedNominationEmptyOrOlder is to check whether the version inside _savedNominatedConfiguredVersion is out of dated, but here we check versions inside configuredInputs is out of dated or not. So, we compare all versions inside configuredInputs with versions inside LoadedProjects.